### ISSUE TITLE
PyTorch Implementation

### ISSUE THREAD

**Issue by crlandsc** [nomonosound/log-wmse-audio-quality#6]
This looks like a very useful audio quality metric, thank you for sharing! However, I am training models using PyTorch, and torch cannot leverage numpy-based operations for backpropagation. Have you (or anyone) tried implementing (or have plans to implement) this metric in torch? Your insights would be greatly appreciated, thanks!

**Comment by iver56** [comment]
Hey :) Thanks for the feedback. I'm glad you think it looks useful.

The current intended use is as a metric, as in evaluating model outputs after training, and not as a differential loss function to be used for model training.

Adapting it for use as a loss function would require some fundamental changes, as it currently inputs three sounds, not two.

I have not made an implementation in pytorch. Would you like to make one?

**Comment by iver56** [comment]
Hmm, after giving it some more thought, maybe fundamental changes are not needed, as we could probably still use it as a loss function even if it inputs three sounds. So maybe it would be viable as a loss function

**Comment by crlandsc** [comment]
That was my thought too. In the description, we would just caveat that it requires the raw audio along with the estimates and actuals (just as you already have in the current implementation).

I see no inherent reason it could not be used as a loss as the frequency weighting (implemented via convolution), log, and MSE should all be supported by torch.

Have you found that it has yielded better results than pinning your progress on SI-SDR (which is pretty standard for a lot of source separation work)? I have been searching for improved losses to train against (L1, L1SNR, RMSE, etc.), but nothing seems to be a clear winner.

I would be potentially interested in implementing a torch version, depending on how useful it might be for source separation.

**Comment by iver56** [comment]
Yeah, I hear you. If making it as a loss function, I would probably want to have it without the resampling. Instead, the frequency weighting could be adapted work with the specific sample rate being used

> Have you found that it has yielded better results than pinning your progress on SI-SDR (which is pretty standard for a lot of source separation work)?

Si-SDR is okay when all the sounds are without digital silence. When there are cases where the model should ideally output digital silence, I have found that I cannot rely on SI-SDR alone.

> I have been searching for improved losses to train against (L1, L1SNR, RMSE, etc.), but nothing seems to be a clear winner.

"Hybrid" losses seem to be all the rage lately. Like L1 (or L1SNR) plus some sort of multi resolution STFT. What is working well for you right now?

**Comment by crlandsc** [comment]
To be totally honest, I have gone back to using L1 for many training runs. I track many other metrics, including multi-res STFT, Sum & Difference (for stereo signals), etc. but I have found that L1 generally seems to train consistently and well enough. I have tried balancing multiple losses as well, but have not had the opportunity to do an effective evaluation between the many possible combinations.

You are using an FIR to do the filtering, correct? I assume that I could just load that in and resample the filter (itself) as needed.
